### PR TITLE
"Battery and bucket" merge back to "timing"

### PR DESCRIPTION
**Pull request by aufdenkampe**
@SRGDamia1 and @bschulz1701, let's start a conversation about what needs to happen to merge the "Battery and bucket" branch back into the "timing" branch, which is functionally serving as the head of Develop.

I would like to complete this merge by Thursday or Friday, so that I can deploy some stations from this merged code next week.

### CODE REVIEW

**Comment by aufdenkampe** [comment]
@SRGDamia1, could you resolve the conflicts?
@bschulz1701, could you clean up the documentation for the new files that you contributed?
Thanks!

**Comment by SRGDamia1** [comment]
I think it is mostly clean-up and documentation that needs to happen.

**Comment by aufdenkampe** [comment]
@SRGDamia1, I'm looking at the diffs and it seems that @bschulz1701 applied Friday's "waitForWarmUp & waitForStability" bug fixes to the `SensorBase.cpp`, where you applied similar fixes to `YosemitechParent.cpp`.  Could you figure out which makes the most sense and align the two branches for a clean merge?

**Comment by SRGDamia1** [comment]
I think "battery" should probably be called "ExternalVoltage" or "ExternalADCVoltage" and it might be good to add a "multiplier" input for using with a voltage divider.  That makes it more clear and more general.

**Comment by SRGDamia1** [comment]
I think "tipping bucket" should also be renamed, because it's not for a direct connection to a bucket.

**Comment by aufdenkampe** [comment]
@SRGDamia1, thanks for your work at resolving conflicts.
I like those suggested renaming changes.
* Change "battery" to "ExternalADCVoltage"
* Change "TippingBucket" to one of the following (I'm looking for votes):
  * "TipCounter"
  * "Counter"
  * "CounterI2C"

I think it would then make sense to rename our https://github.com/EnviroDIY/TippingBucketRainGauge repo to reflect the new name above.

Thoughts?

**Comment by SRGDamia1** [comment]
Also, unrelated to the counter or battery, I'm still not 100% certain I've got the bugs out of the timing branch.  SDI-12 still isn't working for me.

**Comment by aufdenkampe** [comment]
@SRGDamia1, good point in https://github.com/EnviroDIY/ModularSensors/pull/114#issuecomment-380251388 regarding our need to also get depth of rain in English or metric units. 

@bschulz1701, can you add this as user input? Where the modular function allows the user to select water depth per tip, and the units? Then this "rain tip counter" sensor object outputs two variables: tips and rainDepth. Connected to #84 & https://github.com/EnviroDIY/TippingBucketRainGauge.

Regarding library file names, how about changing "TippingBucket" to one of the following (I'm looking for votes):
- "RainTipCounter"
- "RainTipVolumeCounter"
- "RainTipVolumeCounterI2C"
I'm _very_ open to better ideas on this.

**Comment by aufdenkampe** [comment]
Regarding https://github.com/EnviroDIY/ModularSensors/pull/114#issuecomment-380251662, it's good to know (but disappointing to hear) that our Friday bug fixes didn't solve the SDI12 issues. Lucky for me, I don't have SDI12 sensors in this upcoming deployment. That said, @fisherba does have some coming soon (Decagon 5tm soil moisture).

**Comment by aufdenkampe** [comment]
@bschulz1701, let's default to metric, and 0.02 mm per tip.

**Comment by bschulz1701** [comment]
0.02 mm or 0.2 mm? On Davis' website they say the metric version is calibrated to 0.2mm (just want to make sure)

**Comment by aufdenkampe** [comment]
Whoops. Thanks for catching that. It's 2.0 mm for metric, and 0.01" for English.

**Comment by bschulz1701** [comment]
Ok, I will default it to 0.2mm then

By the way, how do we feel about 3 variables, one for number of tips, and one for english measurement and one for metric measurement? 

Initially I was thinking you would just pass the multiplier and the unit of measure as a string, but I'm running into an issue that I can't seem to find a way to call this within the volume variable to pass the units type which was set in the initialization of the sensor `parentSense -> _units`, if that makes sense? 

Unless @SRGDamia1 has the C++ skills to know how to do what I am trying to do here? (Or has a better idea about how to do it?)

**Comment by bschulz1701** [comment]
So I have a solution which just uses the 2 variables, and it is perfectly functional, the initialization is a bit more clunky than I would like, because you have to declare the units both in the function initialization and in the variable initialization, so for the default it is nice and streamlined, you just get:
```
TippingBucket tip(TippingPower);
...
new TippingBucket_Vol(&tip)
```
But if you want to change it to inches, then you have to do this:
```
TippingBucket tip(TippingPower, 0x08, 1, 0.01);
...
new TippingBucket_Vol(&tip, "", "", "in")
```

And actually, as I'm writing this I'm thinking, why are we bothering to represent it in inches? That feature is not present for any other sensors (like to represent the temp as Fahrenheit of Celsius), so why don't we just make the input "mm of rain per tip" and then if you are using a system that tips once every 0.01", you would just need to convert that, input 0.254, and then you will get a readout in mm in the file. 

I think what I just said above makes the most sense and is in keeping with the rest of the sensor systems, so I am going to change it over to that unless anyone has any objections to that

**Comment by bschulz1701** [comment]
As of https://github.com/EnviroDIY/ModularSensors/pull/114/commits/d27a8f964f85d7709eb96cb300757332b251a7ca
**Completed**: 
* updated the documentation on all of the libraries I have added
* changed the name of BatteryMonitor to ExternalVoltage
* implemented the functionality I described in my comment about the tipping bucket

**Incomplete**:
* Changing name of TippingBucket, waiting on decision on what that should be called

**Comment by aufdenkampe** [comment]
@bschulz1701, I like the idea of just going with 2 variables and with metric/mm, and requiring the user to put a different volume per tip.

@SRGDamia1, and suggestions on library and object names, to best conform with the other stuff you have? My last suggestions were to change "TippingBucket" to one of the following (I'm looking for votes):
- "RainTipCounter"
- "RainTipVolumeCounter"
- "RainTipVolumeCounterI2C"

Any suggestions on the default variable and object names?

File: examples/simple_logging/simple_logging.ino [hunk 1]
```diff
@@ -281,6 +281,22 @@ const char *DOptoDI12address = "5";  // The SDI-12 Address of the Zebra Tech D-O
 // const int8_t SDI12Power = 22;  // Pin to switch power on and off (-1 if unconnected)
 ZebraTechDOpto dopto(*DOptoDI12address, SDI12Power, SDI12Data);
 
+#include <ExternalVoltage.h>
+const int8_t VoltPower = 22;  // Pin to switch power on and off (-1 if unconnected)
+const int8_t VoltData = 0;  // The data pin ON THE ADS1115 (NOT the Arduino Pin Number)
+const uint8_t Volt_ADS1115Address = 0x48;  // The I2C address of the ADS1115 ADC
+ExternalVoltage extvolt(VoltPower, VoltData);
+
+#include <TippingBucket.h>
```
Comment (aufdenkampe) [inline examples/simple_logging/simple_logging.ino]: @bschulz1701, this example code looks like it needs updating to include the I2C address.

File: examples/simple_logging/simple_logging.ino [hunk 2]
```diff
@@ -306,13 +322,15 @@ Variable *variableList[] = {
...
     new DecagonES2_Temp(&es2),
+    new ExternalVoltage_Volt(&bat),
     new MaxBotixSonar_Range(&sonar1),
     new MaxBotixSonar_Range(&sonar2),
     new MaximDS18_Temp(&ds18_1),
     new MaximDS18_Temp(&ds18_2),
     new MaximDS18_Temp(&ds18_3),
     new MaximDS18_Temp(&ds18_4),
     new MaximDS18_Temp(&ds18_5),
+    new TippingBucket_Tips(&tip),
```
Comment (aufdenkampe) [inline examples/simple_logging/simple_logging.ino]: This TippingBucket example needs to be updated to 2 variables.

File: examples/simple_logging/simple_logging.ino (hunk 1 above)
Comment (bschulz1701) [reply examples/simple_logging/simple_logging.ino]: Fixed as of https://github.com/EnviroDIY/ModularSensors/pull/114/commits/872fe46629247bfde5cca95d6726f4c6cb7022a5

**Comment by SRGDamia1** [comment]
I think I'm ready to merge this in now.  I did some clean up, made the default gain for 1x for external voltage, because that seemed more general to me, and removed the power pin and averaging options from the tipping bucket, because having those options made no sense to me.

**Comment by aufdenkampe** [comment]
@SRGDamia1 , thank you!

**Comment by SRGDamia1** [comment]
I think RainTipDepthCounterI2C is a huge mouthful.  Maybe just RainTipCounterI2C or ExternalI2CRainCounter?

**Comment by SRGDamia1** [comment]
Changed to RainCounterI2C in e49fa4aab13c65bfea9408094d6517c1752ad780

**Comment by aufdenkampe** [comment]
I like your idea to keep it short, so building/subtracting from your last one, how about "ExternalRainCounter"?

**Comment by aufdenkampe** [comment]
Whoops, I missed your last idea and commit. RainCounterI2C works great.

**Pull request merged**